### ISSUE TITLE
Application placement directive cannot be updated

### ISSUE THREAD

**Issue by gboutry** [juju/terraform-provider-juju#179]
### Terraform Version
Run `terraform -v` to show the version. If you are not running the latest version of Terraform, please upgrade because your issue may have already been fixed.

```bash
$ terraform -v       
Terraform v1.4.1
on linux_amd64
```

```bash
$ juju --version
3.1.0-genericlinux-amd64
```

### Affected Resource(s)
- juju_machine

### Terraform Configuration Files

```hcl
terraform {
  required_providers {
    juju = {
      source  = "terraform.local/juju/juju"
      version = "> 0.5.2"
    }
  }
}

data "juju_model" "control-plane" {
  name = "controller"
}

data "juju_machine" "baremetal" {
  count      = length(var.machine_ids)
  model      = data.juju_model.control-plane.name
  machine_id = var.machine_ids[count.index]
}

resource "juju_application" "microk8s" {
  name        = "microk8s"
  trust       = true
  model       = data.juju_model.control-plane.name
  units       = length(data.juju_machine.baremetal[*])
  placement   = join(",", sort(data.juju_machine.baremetal[*].machine_id))

  charm {
    name     = "microk8s"
    channel  = "latest/stable"
    revision = 28      # revision 28 does not support strict confinement
    series   = "jammy" # Base ?
  }

  config = {
    channel = var.microk8s_channel
    addons  = "dns:8.8.8.8,8.8.4.4 hostpath-storage metallb:10.20.21.1-10.20.21.10"
  }
}
```

with `var.machine_ids=["0", "1", "2"]`

When changing `var.machine_ids` to `["0", "2"]`

Here's the plan
```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # juju_application.microk8s will be updated in-place
  ~ resource "juju_application" "microk8s" {
        id          = "controller:microk8s"
        name        = "microk8s"
      ~ placement   = "0,1,2" -> "0,2"
      ~ units       = 3 -> 2
        # (5 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }
```

### Debug Output
Please provider a link to a GitHub Gist containing the complete debug output: 
https://gist.github.com/gboutry/6ba2ba0bb2a9375685602f3e76b924bc

### Expected Behavior
What should have happened?

Charm's unit on machine 1 should have been removed

### Actual Behavior

Charm's unit was removed on machine 2

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform apply` # with the 3 machines
2. remove any machine for the variable `machine_ids` # juju seems to remove a machine randomly
3. `terraform apply`

### Important Factoids
This has been performed on a manually provisioned cloud with 3 virtual machines.
Terraform juju provider was built with 3b89ac3c13ddb7e264b39dd06c27b2cd3e89ba36

This charm has been provisioned in `controller's` model, which is bad practice, but needed for my current setup. I'm not convinced this has an impact. Juju seems to choose the machine to remove randomly.

(Seems random <-- based on observation, I've reproduced the steps multiple times to make sure, and the removed machine was sometimes 0, sometimes 1 and sometimes 2)

**Comment by gboutry** [comment]
Seems like placement is not updated when changing plan (I don't know if it's possible in juju semantics): https://github.com/juju/terraform-provider-juju/blob/28ee60902a0988b9f1eb68645cd1a3071eb595b4/internal/provider/resource_application.go#L348

**Comment by gboutry** [comment]
After looking at the history, this is more of a feature request than a bug.

**Comment by juanmanuel-tirado** [comment]
You're right. In Juju there are no semantics for a change in the placement directive. From an API perpective, the placement directive can be used during the application creation [here](https://github.com/juju/terraform-provider-juju/blob/28ee60902a0988b9f1eb68645cd1a3071eb595b4/internal/juju/applications.go#L344) but it has to be reconstructed during the reading [here](https://github.com/juju/terraform-provider-juju/blob/28ee60902a0988b9f1eb68645cd1a3071eb595b4/internal/juju/applications.go#L542). The `placement` is not a concept itself inside Juju.

This creates some friction with the way Terraform works. We could emulate and update in the placement directive by running a diff, removing and then redeploying the application to the corresponding machines. However, that logic will only exist on the Terraform provider and we're trying to avoid this.

I think, we will have to think about how to proceed in this case.

**Comment by gboutry** [comment]
Thanks for your response @juanmanuel-tirado, from what I understand, the closest semantics in Juju is to call juju remove-unit / juju add-unit from a reconstructed placement.

Would there be a possibility (in Juju's future), to be able to update Juju's topology directly with a crafted placement directive ? (I agree that it would be much better to have this implemented in the Juju side)

**Comment by juanmanuel-tirado** [comment]
That's not in Juju's short term roadmap that's for sure. We should evaluate this.

Meanwhile, I created a `requires-juju-changed` label and leave this issue open until I can have a further explanation rather than a won't do.

Thanks for rising the issue.

**Comment by SimoneDutto** [comment]
We believe this has been solved in https://github.com/juju/terraform-provider-juju/pull/716

Feel free to re-open this issue or file a new one.
